### PR TITLE
Add Knowledge Graph documentation

### PR DESCRIPTION
**Pull request by jonathanglassman**
Document how the Knowledge Graph works, and what the data sources, data files and data fields are.

### CODE REVIEW

File: source/analysis/govgraph/index.html.md.erb [hunk 1]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
 - upskill colleagues and drive culture change to make better use of data and improve data literacy in GOV.UK
 - provide early adopters with access and training to help scale use of our data apps and product
 
-One of the most common uses is that GOV.UK Knowledge Graph provides a queryable platform for content designers to answer common content questions.
+The Knowledge Graph connects siloed data from analytics (the functional network) to content data metadata such as the title, body, or which organisations publish content. We also use graph algorithms to derive new features, such as the popularity of content with the weighted PageRank algorithm.
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:18]: If you mention the functional network, you might as well mention the structural network: the functional network is built from user movement between pages (obtained through GA) and the structural network (built using the content store) is a graph where the nodes are the gov.uk pages and the edges are hyperlinks between those pages.

File: source/analysis/govgraph/index.html.md.erb [hunk 2]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
 
-There are two versions of the graph:
+The Knowledge Graph has 2 data sources, the Content Store and Google Analytics.
 
-- a [stable graph](https://knowledge-graph.integration.govuk.digital:7473/browser/) for content designers to query
-- an [experimental graph](https://knowledge-graph-lab.integration.govuk.digital:7473/browser/) for the Data Labs team which includes extra NLP derived features
+To create the Knowledge Graph, a [python script](https://github.com/alphagov/govuk-knowledge-graph/blob/master/build_knowledge_graph_data) is run automatically every day at 5am. This script:
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:22]: It's actually a bash script that runs several python programs.

File: source/analysis/govgraph/index.html.md.erb [hunk 3]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
 
-The graph connects siloed data from analytics (the functional network) to content data metadata such as the title, body, or which organisations publish content. We also use graph algorithms to derive new features, such as the popularity of content with the weighted PageRank algorithm.
+- manipulates and transforms data from these data sources
+- saves that transformed data into intermediate data files in AWS S3 buckets
 
-Complete this [introduction lesson](https://docs.google.com/document/d/1R8vati8E5aPJk_p0RYxu2yw2gJK_y6ikvi5VReV4YVs/edit) to understand more about how GOV.UK Knowledge Graph works.
+Another Python script is then run automatically to create the Knowledge Graph from these intermediate data files.
+
+There are 2 Knowledge Graphs, depending on which script is run:
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:29]: I would phrase that as "There are two  instances of the knowledge graph", built separately every morning by running slightly different scripts

File: source/analysis/govgraph/index.html.md.erb [hunk 4]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
+
+- `knowldge_graph`
+- `structural_network`
+- `functional-network/DATE/`
+
+The `govuk-data-infrastructure-integration/knowldge_graph` AWS S3 bucket and sub-directory has the following intermediate data files.
+
+### preprocessed_content_store_<DDMMYY>.csv.gz
+
+A list of all the GOV.UK content pages with many attributes for each page.
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:59]: not all of them, but most

File: source/analysis/govgraph/index.html.md.erb [hunk 5]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
+
+This file is created by the `CreateContentGraph` Python class and has the following data fields.
+
+| Field| Description| Example|
+| --- | ---- | ---- |
+| content_id | the content ID of the page containing the chapter | [](https://www.gov.uk/foreign-travel-advice/french-guiana)[020d58a7-f069-4eb8-bdf2-9979e7196ba4](https://www.gov.uk/foreign-travel-advice/french-guiana) |
+| chapter_number | the index of the chapter on that page| 3|
+
+### chapterid_nodelist.csv
+
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:179]: I think it's worth introducing it with something like: "some pages on www.gov.uk are structured in chapters (<examples>) and the knowledge graph is able to extract them in order to create specific nodes for each chapter."

File: source/analysis/govgraph/index.html.md.erb [hunk 6]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
 - upskill colleagues and drive culture change to make better use of data and improve data literacy in GOV.UK
 - provide early adopters with access and training to help scale use of our data apps and product
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:17]: I'm not sure where I would put this but I think an important goal of the knowledge graph is to:
- merge structural information as well as user behaviour in a single data structure for easy querying and insight gathering
- extract meaning from the content of gov.uk beyond just pages and hyperlinks. We are aiming to build nodes that represent "real-life" concepts such as person, address, eligibility criterion, benefit, etc.

File: source/analysis/govgraph/index.html.md.erb [hunk 7]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
+| --- | ---- | --- |
+| role_content_id | The contentID of the role (does not map to a page) | 8472caef-c0f1-11e4-8223-005056011aef|
+| role_title | The name of the role | Chief Executive |
+| organisation_content_id | the contentID of the organisation| [](https://www.gov.uk/government/organisations/natural-england)[d3ce4ba7-bc75-46b4-89d9-38cb3240376d](https://www.gov.uk/government/organisations/natural-england)|
+| organisation_base_path| The URL of the organisation| [](https://www.gov.uk/government/organisations/natural-england)[/government/organisations/natural-england](https://www.gov.uk/government/organisations/natural-england) |
+| organisation_title | The name of the organisation | Natural England |
+
+## Knowledge Graph creation
+
+Every morning, the [`app-knowledge-graph` bash script](https://github.com/alphagov/govuk-aws/blob/master/terraform/projects/app-knowledge-graph/userdata.tpl) is run automatically to create the Knowledge Graph.
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:396]: Every morning, two applications are started sequentially: `app-data-science` and `app-knowledge-graph` [link to the respective folders in govuk-aws]. 

`app-data-science` runs the `build_knowledge_graph_data` which generates all the files described above and copies them to S3

`app-knowledge-graph` generates runs the bash scripts `provision_knowledge_graph` and `provision_knowledge_graph_lab`, which build the two knowledge graphs from the files in S3

File: source/analysis/govgraph/index.html.md.erb [hunk 8]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
+- the [user-facing Knowledge Graph](https://knowledge-graph.integration.govuk.digital:7473/browser/) that content designers use to answer common content questions
+- the [more advanced Knowledge Graph](https://knowledge-graph-lab.integration.govuk.digital:7473/browser/) that data professionals use for experiments
+
+## Further information on the Knowledge Graph
 
 For more information, see the following:
 
+- the [introduction lesson on the Knowledge Graph](https://docs.google.com/document/d/1R8vati8E5aPJk_p0RYxu2yw2gJK_y6ikvi5VReV4YVs/edit) to understand more about how it works
 - this [blog post on Knowledge Graph](https://insidegovuk.blog.gov.uk/2020/08/07/one-graph-to-rule-them-all/)
 - the [GOV.UK developer documentation Knowledge Graph page](https://docs.publishing.service.gov.uk/manual/knowledge-graph.html)
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:416]: maybe the page on the dev docs should be edited to just point to this page.

File: source/analysis/govgraph/index.html.md.erb [hunk 9]
```diff
@@ -15,19 +15,403 @@ The [GOV.UK Knowledge Graph](https://github.com/alphagov/govuk-knowledge-graph),
...
 - [example queries](https://github.com/alphagov/govuk-knowledge-graph/tree/master/src/neo4j/favourite_queries) in the repo
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:417]: Is this documentation going to be public (like the developer docs), btw. Are all the links and content ok to be so? Just wondering.

**Review by maxf — CHANGES_REQUESTED** [review]

File: source/analysis/govgraph/index.html.md.erb (hunk 8 above)
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:416]: agree, let's remove this link for now as this has far more information

File: source/analysis/govgraph/index.html.md.erb (hunk 9 above)
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:417]: They are public yes. Are there any links that we should remove?

Comment (maxf) [reply source/analysis/govgraph/index.html.md.erb:405]: ok, so:
- the first link isn't public but I suppose it's ok to link to it anyway
- Shouldn't the dev docs page on the KG be incorporated to this doc and removed eventually?
- The example queries aren't actual examples, they're just the cypher commands to build the graph from the data. If you want to show sample queries, you probably need to introduce neo4j and the cypher language. Exactly what the document in the first link describes, in fact.

File: source/analysis/govgraph/index.html.md.erb (hunk 8 above)
Comment (maxf) [reply source/analysis/govgraph/index.html.md.erb]: ok

File: source/analysis/govgraph/index.html.md.erb (hunk 5 above)
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:182]: agree and done

File: source/analysis/govgraph/index.html.md.erb [hunk 10]
```diff
@@ -1,37 +1,406 @@
...
+| role_appointment_content_id | contentID of the appointment (does not have a corresponding page) | 9f0833ef-3faa-46e5-a7c5-f6ca841398c6 |
+| role_appointment_title | Name of the person and role | Rena Lalgie - Governor to Bermuda|
+| role_appointment_started_on | data when role started| 2020-12-10T00:00:00+00:00|
+| role_appointment_ended_on | date when role ended (if applicable)||
+| role_content_id| contentID of the role (no corresponding page) | 847816b4-c0f1-11e4-8223-005056011aef |
+| role_title| Name of the role| Governor to Bermuda|
+
+### roles_nodelist.csv
+
+A list of all the roles.
```
Comment (maxf) [inline source/analysis/govgraph/index.html.md.erb:378]: same again, what's a role, etc.

File: source/analysis/govgraph/index.html.md.erb [hunk 11]
```diff
@@ -1,37 +1,406 @@
...
+
+To create the Knowledge Graph, every day at 5am the following applications are run in sequence:
+
+- [`app-data-science`](https://github.com/alphagov/govuk-aws/tree/master/terraform/projects/app-data-science)
+- [`app-knowledge-graph`](https://github.com/alphagov/govuk-aws/blob/master/terraform/projects/app-knowledge-graph)
+
+The `app-data-science` application runs the [`build_knowledge_graph_data` Python script](https://github.com/alphagov/govuk-knowledge-graph/blob/master/build_knowledge_graph_data) which generates [intermediate data files](#intermediate-data-files) from the 2 data sources and copies those files to the following AWS S3 buckets:
+
+- `govuk-data-infrastructure-integration` in the following subdirectories:
+    - `knowldge_graph`
```
[the remark this answers is not in the document]
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:44]: I think I got that from the spreadsheet, so is there a typo in the spreadsheet or does the sub-dir in the AWS S3 bucket also have this typo?

File: source/analysis/govgraph/index.html.md.erb [hunk 12]
```diff
@@ -1,37 +1,406 @@
...
+
+| Field| Description| Example|
+| ---- | ---- | ---- |
+| content_id| The content_id of a page| [](https://www.gov.uk/employment-tribunal-decisions/mrs-c-harris-v-worldaware-ltd-2206826-2018)[00000700-262d-441f-8850-cbce15addfb4](https://www.gov.uk/employment-tribunal-decisions/mrs-c-harris-v-worldaware-ltd-2206826-2018) |
+| orgid| The GA code for the primary publishing organisation | EA73 |
+| org_content_id | The content_id of the page describing a primary publishing organisation | [](https://www.gov.uk/government/organisations/hm-courts-and-tribunals-service)[6f757605-ab8f-4b62-84e4-99f79cf085c2](https://www.gov.uk/government/organisations/hm-courts-and-tribunals-service) |
+
+### cid_has_supporting_organisations_edgelist.csv
+
+A list of content pages with all supporting organisations.
```
[the remark this answers is not in the document]
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:149]: agree

File: source/analysis/govgraph/index.html.md.erb [hunk 13]
```diff
@@ -1,37 +1,406 @@
...
+| Field| Description| Example|
+| --- | --- | ---- |
+| src_orgid| The analytics ID of the superseding organisations | OT1211 |
+| src_title| The name of the superseding organisation| National Cyber Security Centre |
+| target_orgid | the analytics ID of the superseded organisation | OT563|
+| target_title | The name of the superseded organisation | CESG |
+
+### people_nodelist.csv
+
+A list of pages about people.
```
[the remark this answers is not in the document]
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:348]: not sure either

**Review by maxf — APPROVED** [review]
:+1: 
Thanks for all the fixes

File: source/analysis/govgraph/index.html.md.erb [hunk 14]
```diff
@@ -1,37 +1,435 @@
...
+
+The following intermediate data files are used to build the 2 instances of the Knowledge Graph.
+
+### preprocessed_content_store_<DDMMYY>.csv.gz
+
+A list of most GOV.UK content pages with multiple attributes for each page.
+
+The `src.data.extract_contect_store.py` Python script creates a [pandas dataframe](https://pandas.pydata.org/docs/reference/frame.html) and exports a copy of this dataframe to the S3 bucket and sub-directory as the `.csv` file.
+
+This script also creates the `CreateContentGraph` Python class which other intermediate data files are created from.
```
Comment (exfalsoquodlibet) [inline source/analysis/govgraph/index.html.md.erb:73]: Not very clear to me, I suggest we replace it with the below (please rephrase as you think best)

```suggestion
The `ContentStore` Python class, defined in [`src.data.extract_contect_store.py`](https://github.com/alphagov/govuk-knowledge-graph/blob/master/src/data/extract_content_store.py) and instantiated in [`src.build_graph_data.py`](https://github.com/alphagov/govuk-knowledge-graph/blob/master/src/build_graph_data.py) retrieves the relevant content-store data and saves it to a [pandas dataframe](https://pandas.pydata.org/docs/reference/frame.html). 
A copy of this dataframe is then exported to the `govuk-data-infrastructure-integration/knowldge_graph/` subdirectory in the AWS S3 bucket by the [`build_knowledge_graph_data`](https://github.com/alphagov/govuk-knowledge-graph/blob/master/build_knowledge_graph_data) script as `preprocessed_content_store_<DDMMYY>.csv.gz` file.

The dataframe is also used to instantiate the [`CreateContentGraph` Python class](https://github.com/alphagov/govuk-knowledge-graph/blob/master/src/data/create_graph.py), which then generates other intermediate data files, further components of the Knowledge Graph (see below).
```

File: source/analysis/govgraph/index.html.md.erb [hunk 15]
```diff
@@ -1,37 +1,435 @@
...
+
+Once you have the username and password, you can view the updated [user-facing Knowledge Graph](https://knowledge-graph.integration.govuk.digital:7473/browser/) and the [experimental Knowledge Graph](https://knowledge-graph-lab.integration.govuk.digital:7473/browser/) for data professionals.
+
+## Intermediate data files
+
+The following intermediate data files are used to build the 2 instances of the Knowledge Graph.
+
+### preprocessed_content_store_<DDMMYY>.csv.gz
+
+A list of most GOV.UK content pages with multiple attributes for each page.
```
Comment (exfalsoquodlibet) [inline source/analysis/govgraph/index.html.md.erb:69]: It is a dataset more than a list.

File: source/analysis/govgraph/index.html.md.erb [hunk 16]
```diff
@@ -1,37 +1,435 @@
...
+
+The `src.data.extract_contect_store.py` Python script creates a [pandas dataframe](https://pandas.pydata.org/docs/reference/frame.html) and exports a copy of this dataframe to the S3 bucket and sub-directory as the `.csv` file.
+
+This script also creates the `CreateContentGraph` Python class which other intermediate data files are created from.
+
+This file:
+
+- excludes [document types that are blocked from the topic taxonomy](https://github.com/alphagov/govuk-knowledge-graph/blob/master/config/document_types_excluded_from_the_topic_taxonomy.yml)
+- only retrieves data for documents with a `Locale` field of `English`
+- only retrieves content items with a `Phase` field of `Live`
```
Comment (exfalsoquodlibet) [inline source/analysis/govgraph/index.html.md.erb:79]:
```suggestion
This dataset is made of Content Store data, but:

- excluding [document types that are blocked from the topic taxonomy](https://github.com/alphagov/govuk-knowledge-graph/blob/master/config/document_types_excluded_from_the_topic_taxonomy.yml)
- only retrieving data for documents with a `Locale` field of `English`
- only retrieving content items with a `Phase` field of `Live`
```

File: source/analysis/govgraph/index.html.md.erb [hunk 17]
```diff
@@ -1,37 +1,435 @@
...
+
+A lookup table to find the content ID of a page from its URL, or the URL of a page from its content ID.
+
+This file is created by the `src.data.extract_contect_store.py` Python script.
+
+An example data field is `"/coronavirus-health-worker-visa-extension": "59ecfc6d-cb9e-49b8-b013-fc368ebdf8ed"`.
+
+### cid_has_organisations_edgelist.csv
+
+A list of content pages with all organisations for those pages.
```
Comment (exfalsoquodlibet) [inline source/analysis/govgraph/index.html.md.erb:118]: I know we have been using "pages" throughout. As long as we defined what we mean we page, this is fine. It is a content item, not a url.

File: source/analysis/govgraph/index.html.md.erb [hunk 18]
```diff
@@ -1,37 +1,435 @@
...
+The `CreateContentGraph` Python class creates this file, which has the following data fields.
+
+| field | description| example|
+| ---- | --- | ---- |
+| content_id| Content ID of a page | [00022740-0ba7-4fd0-8ca8-f0c3d6156fec](https://www.gov.uk/research-for-development-outputs/information-sharing-credit-booms-and-financial-stability-lessons-for-developing-economies) |
+| taxon_content_id | Content ID of the page’s taxon | [9fb30a53-70fb-4f1c-878b-0064b202d1ba](https://www.gov.uk/international/international-aid-and-development) |
+
+### cid_has_chapterid_edgelist.csv
+
+Some GOV.UK pages are divided into chapters. The Knowledge Graph’s build process extracts these chapters as separate data points and presents them as nodes in the Knowledge Graph.
```
Comment (exfalsoquodlibet) [inline source/analysis/govgraph/index.html.md.erb:185]: We could add an example?

Here is one of such pages: https://www.gov.uk/universal-credit

File: source/analysis/govgraph/index.html.md.erb [hunk 19]
```diff
@@ -1,37 +1,435 @@
...
+This file is created by the [`extract_publishing_api.py` Python script](https://github.com/alphagov/govuk-knowledge-graph/blob/master/src/data/extract_publishing_api.py) and has the following data fields.
 
-Complete this [introduction lesson](https://docs.google.com/document/d/1R8vati8E5aPJk_p0RYxu2yw2gJK_y6ikvi5VReV4YVs/edit) to understand more about how GOV.UK Knowledge Graph works.
+| field| description | example |
+| --- | ---- | --- |
+| role_content_id | Content ID of the role (does not map to a page) | 8472caef-c0f1-11e4-8223-005056011aef|
+| role_title | Name of the role | Chief Executive |
+| organisation_content_id | Content ID of the organisation| [d3ce4ba7-bc75-46b4-89d9-38cb3240376d](https://www.gov.uk/government/organisations/natural-england)|
+| organisation_base_path| URL of the organisation| [/government/organisations/natural-england](https://www.gov.uk/government/organisations/natural-england) |
+| organisation_title | Name of the organisation | Natural England |
```
Comment (exfalsoquodlibet) [inline source/analysis/govgraph/index.html.md.erb:429]: I feel like something is missing here.

Could we add what Labels and Relationships are built using these intermediate files? We are not saying anything about how the knowledge graph is actually built, i.e., how all these bits and pieces come together.
Like, what happen in the `provision_knowledge_graph`? What nodes and relationships are created exactly? We need this info here. cc: @maxf 

We are also missing information about `eligibility_pages` and `external_links` intermediate files which are also created as part of building the graph data.

Can we also add an image/screenshot of the knowledge graph?

**Review by exfalsoquodlibet — COMMENTED** [review]
Wow, lots of work @jonathanglassman! Great stuff!

Some comments and questions from me. Happy to discuss of course. Thanks!

File: source/analysis/govgraph/index.html.md.erb [hunk 20]
```diff
@@ -1,37 +1,435 @@
...
+## Data sources
+
+The Knowledge Graph has 2 data sources:
+
+- the [Content Store](https://github.com/alphagov/content-store)
+- Google Analytics
+
+The Content Store is used to build a structural network. In this network, the network nodes are the GOV.UK pages, and the network edges are hyperlinks between those pages.
+
+Google Analytics data is used to build a functional network. This network connects content data to content metadata such as title or publishing organisation.
```
[the remark this answers is not in the document]
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:32]: From the style guide: "Data - Treat as a singular noun e.g. The data is stored on a secure server."

File: source/analysis/govgraph/index.html.md.erb (hunk 19 above)
Comment (jonathanglassman) [reply source/analysis/govgraph/index.html.md.erb:429]: @exfalsoquodlibet thanks for the review - this change sounds more substantial and we should have a chat

**Pull request merged**